docs(module-resolver): Update README documentation (#1967)
salesforce/lwc: packages/@lwc/module-resolver/README.md

### `resolveModule(specifier, importer, options)`

> Synchronously resolve an LWC module from its specifier from an importer path.

Reviewer: Instead of "resolves" can we use another word or provide a sentence saying what "resolving" is? Is it locating the file and assigning it to the specifier?

console.log(result);
```

> If in the middle of the module resolution the resolver processes an invalid configuration the resolver throws an error with the `LWC_CONFIG_ERROR` error code. If a module can't be resolved from the given importer the invocation throws an error with the `NO_LWC_MODULE_FOUND` error code.

Reviewer: ```suggestion
If the resolver processes an invalid configuration, it throws an error with the `LWC_CONFIG_ERROR` error code. If the resolver can't location the module, it throws an error with the `NO_LWC_MODULE_FOUND` error code.
```

```js
import { resolveModule } from '@lwc/module-resolver';

let result;
try {
    result = resolveModule('x/foo', './index.js');
} catch (err) {
    if (err.code === 'LWC_CONFIG_ERROR') {
        console.error(`The request module can't be resolved due to an invalid configuration`, err);
    } else if (err.code === 'NO_LWC_MODULE_FOUND') {

>         console.error(`The requested module doesn't exists`, error);

Reviewer: ```suggestion
        console.error(`The requested module doesn't exist. Check the directory path.`, error);
```

    } else {
        throw err;
    }
}

console.log(result);
```

**Parameters:**

> -   `specifier` (string, required): The module specifier that should be resolved.

Reviewer: ```suggestion
-   `specifier` (string, required): The module specifier to resolve.
```

> -   `dirname` (string, required): The directory from where the specifier should be resolved.

Reviewer: ```suggestion
-   `dirname` (string, required): The module specifier is resolved from this directory.
```

Author: I don't think we should use the active voice to describe this argument.

-   `options` (object, optional):

>     -   `modules` (ModuleRecord[], optional, default: `[]`): Injects module records to the resolved resolved configuration.

Reviewer: ```suggestion
    -   `modules` (ModuleRecord[], optional, default: `[]`): Injects module records to the resolved configuration.
```

>     -   `rootDir` (string, optional, default: `process.cwd()`): Should only be used when the `modules` option is set. Defines from where the modules overrides should be resolved.

Reviewer: ```suggestion
    -   `rootDir` (string, optional, default: `process.cwd()`): Use only when the `modules` option is set. Modules overrides are resolved from this directory.
```

**Return value:**

A `RegistryEntry` representing the resolved module with the following properties:

-   `entry` (string): The absolute path of the module entry point.
-   `specifier` (string): The resolved module specifier.
-   `scope` (string): The absolute path from where the module has been resolved.

## Module resolution

> The LWC compiler does use a custom resolution algorithm to resolve LWC modules. The module resolution can be configured via the `lwc.config.json` or the `lwc` key on the `package.json`. The `modules` key accepts an array of module records. The module resolver iterates through the `modules` list and returns the first module matching the requested module specifier. A module specifier can be of three different types:

Reviewer: ```suggestion
The LWC compiler uses a custom resolution algorithm to resolve LWC modules. To configure module resolution, use the `lwc.config.json` file or the `lwc` key in the `package.json` file. The `modules` key accepts an array of module records. The resolver iterates through the `modules` array and returns the first module that matches the requested module specifier. There are three types of module record:
```

> -   [Alias module record](#alias-module-record): Define a file path where an LWC module can be resolved.

Reviewer: ```suggestion
-   [Alias module record](#alias-module-record): A file path where an LWC module can be resolved.
```

> -   [Directory module record](#directory-module-record): Defines a folder path where LWC modules can be resolved.

Reviewer: ```suggestion
-   [Directory module record](#directory-module-record): A folder path where LWC modules can be resolved.
```

> -   [NPM package module record](#npm-package-module-record): Defines an NPM package that will expose one or more LWC modules.

Reviewer: ```suggestion
-   [NPM package module record](#npm-package-module-record): An NPM package that exposes one or more LWC modules.
```

```json
// lwc.config.json
{
    "modules": [
        {
            "name": "ui/button",
            "path": "src/modules/ui/button/button.js"
        },
        {
            "dir": "src/modules"
        },
        {
            "npm": "@ui/components"
        }
    ]
}
```

### Module record types

#### Alias module record

An alias module record maps a module specifier to a file path. An alias module record is defined by two keys:

-   `name` (string, required): The LWC module specifier.

> -   `path` (string, required): The path file path to resolve.

Reviewer: ```suggestion
-   `path` (string, required): The file path to resolve.
```

> In this example, the `ui/button` LWC module specifier can be resolved from the `src/modules/ui/button/button.js` path.

Reviewer: ```suggestion
In this example, the `ui/button` LWC module specifier is resolved from the `src/modules/ui/button/button.js` path.
```

```json
{
    "modules": [
        {
            "name": "ui/button",
            "path": "src/modules/ui/button/button.js"
        }
    ]
}
```

#### Directory module record

> A directory module record specifies a folder path where LWC modules can be resolved from. A directory module record is defined by one key:

Reviewer: ```suggestion
A directory module record specifies a folder path where LWC modules are resolved. A directory module record is defined by one key:
```

-   `dir` (string, required): The directory path containing the modules.

```json
{
    "modules": [
        {
            "dir": "src/modules"
        }
    ]
}
```

> The directory module record following uses an opinionated folder structure to resolve LWC modules. The directory path can contain one or multiple folders representing the LWC modules `namespace`. Each of those namespace folders can contain one or multiple directories representing the different LWC modules part of this namespace. The name of the directory indicates the LWC module `name`. For a module to be resolved, the LWC module folder should have a file matching the LWC module name, this file is the LWC module entry point.

Reviewer 2: ```suggestion
The directory module record uses an opinionated folder structure to resolve LWC modules. The directory path can contain one or multiple folders representing the LWC modules `namespace`. Each of those namespace folders can contain one or multiple folders representing the different LWC modules in the namespace. The name of the folder defines the LWC module `name`. For a module to be resolved, the LWC module folder must have a file matching the LWC module name, this file is the LWC module entry point.
```

In this example, if the `dir` key is set `src/modules`, the following LWC modules can be resolved: `ui/button`, `ui/icons`, `shared/utils`.

```
src
└── modules/
    ├── ui/
    │   ├── button/
    │   │   ├── button.js
    │   │   └── button.html
    │   └── icon/
    │       └── icon.js
    └── shared/
        └── utils/
            └── utils.js
```

#### NPM package module record

> An NPM package module record tells the resolver than a given NPM package exposes LWC modules than can be resolved. An NPM package module record is defined by one key:

Reviewer 2: ```suggestion
An NPM package module record tells the resolver that a given NPM package exposes resolvable LWC modules. An NPM package module record is defined by one key:
```

-   `npm` (string, required): The NPM package name exposing the LWC modules.

> In this example, the resolver is told to look into the `@ui/components` npm package to look up for LWC modules.

Reviewer 2: ```suggestion
In this example, the resolver is told to look into the `@ui/components` npm package to look up LWC modules.
```

Reviewer 2: This directive requires the NPM package to follow the _Module resolution_ configuration specified above, right? That is, the NPM package needs either an `lwc.config.json` (at the root of the package, right?) or the `lwc` key in its `package.json`.

Plus it needs to follow the _Exposing LWC modules via NPM packages_ guidance below, right?

Assuming the above is correct, this section needs clarification to make it more obvious how things work.

Author: I added more details.

```json
{
    "modules": [
        {
            "npm": "@ui/components"
        }
    ]
}
```

### Exposing LWC modules via NPM packages

> By default no LWC modules is exposed outside NPM packages. In order to distribute LWC modules via NPM packages, the LWC configuration should list all the public LWC modules on the `expose` key.

Reviewer 2: ```suggestion
By default no LWC module is exposed outside NPM packages. In order to distribute LWC modules via NPM packages, the LWC configuration should list all the public LWC modules on the `expose` key.
```

Reviewer: ```suggestion
By default, an LWC module is not exposed outside of an NPM package. To distribute an LWC module publicly via an NPM package, the LWC configuration must list the LWC module on the `expose` key.
```

In this example, if the `dir` key is set `src/modules`, the following LWC modules can be resolved: `ui/button`, `ui/icons`, `shared/utils`.

```
src
└── modules/
    ├── ui/
    │   ├── button/
    │   │   ├── button.js
    │   │   └── button.html
    │   └── icon/
    │       └── icon.js
    └── shared/
        └── utils/
            └── utils.js
```

#### NPM package module record

An NPM package module record tells the resolver that a given NPM package exposes resolvable LWC modules. More details about how to expose LWC modules out of an NPM package can be found in [this section](#exposing-lwc-modules-via-npm-packages). An NPM package module record is defined by one key:

-   `npm` (string, required): The NPM package name exposing the LWC modules.

> In this example, the resolver is told to look into the `@ui/components` npm package to look up LWC modules.

Reviewer: ```suggestion
In this example, the resolver is told to look into the `@ui/components` NPM package to look up LWC modules.
```

```json
{
    "modules": [
        {
            "npm": "@ui/components"
        }
    ]
}
```

### Exposing LWC modules via NPM packages

To distribute an LWC module publicly via an NPM package, the package should follow the same the [LWC module resolution](#module-resolution) rules. The NPM package should either have the `lwc.config.json` file in its root directory or the `lwc` key in its `package.json` describing how LWC modules are resolved relative to this package.

> By default, an LWC module is not exposed outside of an NPM package. The LWC configuration must list explicitly the public LWC modules on the `expose` key.

Reviewer: ```suggestion
By default, an LWC module is not exposed outside of an NPM package. The LWC configuration must explicitly list the public LWC modules on the `expose` key.
```
